File: Release/WindowsPhone7/PhoneToolkitSample/Data/AllPeople.cs
// (c) Copyright Microsoft Corporation.
// This source is subject to the Microsoft Public License (Ms-PL).
// Please see http://go.microsoft.com/fwlink/?LinkID=131993 for details.
// All other rights reserved.

using System.Collections.Generic;

namespace PhoneToolkitSample.Data
{
    public class AllPeople : IEnumerable<Person>
    {
        private static Dictionary<int, Person> _personLookup;
        private static AllPeople _instance;

        public static AllPeople Current
        {
            get
            {
                return _instance ?? (_instance = new AllPeople());
            }
        }

        public Person this[int index]
        {
            get
            {
                Person person;
                _personLookup.TryGetValue(index, out person);
                return person;
            }
        }

        #region IEnumerable<Person> Members

        public IEnumerator<Person> GetEnumerator()
        {
            EnsureData();
            return _personLookup.Values.GetEnumerator();
        }

        #endregion

        #region IEnumerable Members

        System.Collections.IEnumerator System.Collections.IEnumerable.GetEnumerator()
        {
            EnsureData();
            return _personLookup.Values.GetEnumerator();
        }

        #endregion

        private void EnsureData()
        {
            if (_personLookup == null)
            {
                _personLookup = new Dictionary<int, Person>();
                for (int n = 0; n < 100; ++n)
                {
                    Person person = Person.GetRandomPerson(n);
                    _personLookup[n] = person;
                }
            }
        }

    }
}

File: Release/WindowsPhone7/PhoneToolkitSample/Data/AllPeople.vb
' (c) Copyright Microsoft Corporation.
' This source is subject to the Microsoft Public License (Ms-PL).
' Please see http://go.microsoft.com/fwlink/?LinkID=131993 for details.
' All other rights reserved.

Namespace Data

    Public Class AllPeople
        Implements IEnumerable(Of Person)
        Private Shared _personLookup As Dictionary(Of Integer, Person)
        Private Shared _instance As AllPeople

        Public Shared ReadOnly Property Current As AllPeople
            Get
                If _instance IsNot Nothing Then
                    Return _instance
                Else
                    _instance = New AllPeople
                    Return _instance
                End If
            End Get
        End Property

        Default Public ReadOnly Property Item(ByVal index As Integer) As Person
            Get
                Dim person As Person = Nothing
                _personLookup.TryGetValue(index, person)
                Return person
            End Get
        End Property

#Region "IEnumerable<Person> Members"

        Public Function GetEnumerator() As IEnumerator(Of Person) Implements IEnumerable(Of Person).GetEnumerator
            EnsureData()
            Return _personLookup.Values.GetEnumerator()
        End Function

#End Region

#Region "IEnumerable Members"

        Private Function IEnumerable_GetEnumerator() As System.Collections.IEnumerator Implements System.Collections.IEnumerable.GetEnumerator
            EnsureData()
            Return _personLookup.Values.GetEnumerator()
        End Function

#End Region

        Private Sub EnsureData()
            If _personLookup Is Nothing Then
                _personLookup = New Dictionary(Of Integer, Person)
                For n = 0 To 99
                    Dim person As Person = person.GetRandomPerson(n)
                    _personLookup(n) = person
                Next n
            End If
        End Sub

    End Class

End Namespace
